src/rust/src/rerr.rs

```diff
@@ -0,0 +1,130 @@
+use extendr_api::{extendr, extendr_module, symbol::class_symbol, Attributes, Rinternals, Robj};
+
+#[derive(Clone, Debug)]
+pub enum Rctx {
+    Extendr(String),
+    Hint(String),
+    Plain(String),
+    Polars(String),
+    When(String),
+    TypeMismatch(String, String, String),
+}
+
+#[derive(Clone, Debug)]
+pub struct Rerr(Vec<Rctx>);
+pub type RResult<T> = core::result::Result<T, Rerr>;
+
+pub trait WithRctx<T> {
+    fn ctx(self, rctx: Rctx) -> RResult<T>;
+    fn hint<S: Into<String>>(self, msg: S) -> RResult<T>;
+    fn when<S: Into<String>>(self, msg: S) -> RResult<T>;
+}
+
+impl<T: Clone, E: Into<Rerr>> WithRctx<T> for core::result::Result<T, E> {
```

Comment:
I think I would remove the `: Clone` from `T: Clone`. I think it is better to move T to avoid overhead.

Reply:
I forgot to remove it when I refactor the code. This trait is not needed for us.